pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1905,7 +1902,33 @@ impl RemoteTimelineClient {
                         .await
                         .map_err(|e| anyhow::anyhow!(e))
                 }
-                unexpected @ UploadOp::Barrier(_) | unexpected @ UploadOp::Shutdown => {
+                UploadOp::Barrier(sender) => {
+                    // Barrier operation needs to wait until all deletions are uploaded.
+                    //
+                    // There is an optimization in the deletion queue client that if generations are enabled for a tenant,
+                    // we do not wait for the deletions to be flushed to the remote storage before we tell the user that
+                    // the deletion is complete. This would cause problems for barrier semantics, as we expect all
+                    // deletion to be persistent when the barrier is processed. For example, when we reload a timeline
+                    // without restarting the pageserver (i.e., offload, detach/attach, detach_ancestor), the generation
+                    // number is not bumped.
+                    //
+                    // Let's think about the following operation sequence:
+                    //
+                    // schedule deletion of layer A at generation 5, this can be for example a future layer
```

Comment:
Can we think of any cases other than the "future layer" deletion case that would hit this?  (off the top of my head, I can't)